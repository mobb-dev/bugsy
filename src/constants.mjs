import path from 'node:path';
import { fileURLToPath } from 'node:url';
import * as dotenv from 'dotenv';
import { z } from 'zod';
import Debug from 'debug';

const debug = Debug('mobbdev:constants');
const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, '../.env') });

export const SCANNERS = {
    Checkmarx: 'checkmarx',
    Codeql: 'codeql',
    Fortify: 'fortify',
    Snyk: 'snyk',
};

const envVariablesSchema = z
    .object({
        WEB_LOGIN_URL: z.string(),
        WEB_APP_URL: z.string(),
        API_URL: z.string(),
    })
    .required();

const envVariables = envVariablesSchema.parse(process.env);
debug('config %o', envVariables);

export const mobbAscii = `
                                   ..                       
                             ..........                     
                        .................                   
               ...........................                  
              ..............................                
             ................................               
             ..................................             
            ....................................            
            .....................................           
            .............................................   
          ................................................. 
     ...............................       .................
  ..................................           ............ 
..................    .............            ..........   
......... ........      .........             ......        
  ...............                          ....             
         .... ..                                            
                                                            
                                      . ...                 
                              ..............                
                       ......................               
                   ...........................              
               ................................             
           ......................................           
                ...............................             
                       .................                    
`;

export const WEB_LOGIN_URL = envVariables.WEB_LOGIN_URL;
export const WEB_APP_URL = envVariables.WEB_APP_URL;
export const API_URL = envVariables.API_URL;
