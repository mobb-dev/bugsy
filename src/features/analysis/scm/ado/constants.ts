import { scmCloudUrl } from '../types'

export const ADO_ACCESS_TOKEN_URL =
  'https://app.vssps.visualstudio.com/oauth2/token'

export const DEFUALT_ADO_ORIGIN = scmCloudUrl.Ado

export enum AdoTokenTypeEnum {
  PAT = 'PAT',
  OAUTH = 'OAUTH',
  NONE = 'NONE',
}
