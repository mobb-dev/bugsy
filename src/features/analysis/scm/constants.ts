export const MOBB_ICON_IMG =
  'https://app.mobb.ai/gh-action/Logo_Rounded_Icon.svg'
