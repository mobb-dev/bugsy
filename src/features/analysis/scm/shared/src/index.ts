export * from './commitDescriptionMarkup'
export * from './fixDetailsData'
export * from './getIssueType'
export * from './guidances'
export * from './types'
export * from './urlParser'
export * from './utils'
export * from './validations'
