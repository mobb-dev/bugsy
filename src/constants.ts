import path from 'node:path'
import { fileURLToPath } from 'node:url'

import chalk from 'chalk'
import Debug from 'debug'
import * as dotenv from 'dotenv'
import { z } from 'zod'

import { ScmType } from './features/analysis/scm/shared/src/types'

const debug = Debug('mobbdev:constants')
const __dirname = path.dirname(fileURLToPath(import.meta.url))
dotenv.config({ path: path.join(__dirname, '../.env') })

export const scmFriendlyText: Record<ScmType, string> = {
  [ScmType.Ado]: 'Azure DevOps',
  [ScmType.Bitbucket]: 'Bitbucket',
  [ScmType.GitHub]: 'GitGub',
  [ScmType.GitLab]: 'GitLab',
}

export const SCANNERS = {
  Checkmarx: 'checkmarx',
  Codeql: 'codeql',
  Fortify: 'fortify',
  Snyk: 'snyk',
  Sonarqube: 'sonarqube',
} as const

export const SupportedScannersZ = z.enum([SCANNERS.Checkmarx, SCANNERS.Snyk])
export type SupportedScanners = z.infer<typeof SupportedScannersZ>

type ScannerType = typeof SCANNERS
export type Scanner = ScannerType[keyof typeof SCANNERS]

const envVariablesSchema = z
  .object({
    WEB_APP_URL: z.string(),
    API_URL: z.string(),
    HASURA_ACCESS_KEY: z.string(),
    LOCAL_GRAPHQL_ENDPOINT: z.string(),
  })
  .required()

const envVariables = envVariablesSchema.parse(process.env)
debug('config %o', envVariables)

export const mobbAscii = `
                                   ..                       
                             ..........                     
                        .................                   
               ...........................                  
              ..............................                
             ................................               
             ..................................             
            ....................................            
            .....................................           
            .............................................   
          ................................................. 
     ...............................       .................
  ..................................           ............ 
..................    .............            ..........   
......... ........      .........             ......        
  ...............                          ....             
         .... ..                                            
                                                            
                                      . ...                 
                              ..............                
                       ......................               
                   ...........................              
               ................................             
           ......................................           
                ...............................             
                       .................                    
`

export const PROJECT_DEFAULT_NAME = 'My first project'
export const WEB_APP_URL = envVariables.WEB_APP_URL
export const API_URL = envVariables.API_URL
export const HASURA_ACCESS_KEY = envVariables.HASURA_ACCESS_KEY
export const LOCAL_GRAPHQL_ENDPOINT = envVariables.LOCAL_GRAPHQL_ENDPOINT

export const PROJECT_PAGE_REGEX =
  /^http:\/\/(127\.0\.0\.1)|(localhost):5173\/organization\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\/project\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/

export const errorMessages = {
  missingCxProjectName: `project name ${chalk.bold(
    '(--cx-project-name)'
  )} is needed if you're using checkmarx`,
  missingUrl: `url ${chalk.bold(
    '(--url)'
  )} is needed if you're adding an SCM token`,
  invalidScmType: `SCM type ${chalk.bold(
    '(--scm-type)'
  )} is invalid, please use one of: ${Object.values(ScmType).join(', ')}`,
  missingToken: `SCM token ${chalk.bold(
    '(--token)'
  )} is needed if you're adding an SCM token`,
} as const

export const progressMassages = {
  processingVulnerabilityReportSuccess:
    '⚙️  Vulnerability report proccessed successfully',
  processingVulnerabilityReport: '⚙️  Proccessing vulnerability report',
  processingVulnerabilityReportFailed:
    '⚙️  Error Proccessing vulnerability report',
} as const

export const VUL_REPORT_DIGEST_TIMEOUT_MS = 1000 * 60 * 20 // 20 minutes in msec
