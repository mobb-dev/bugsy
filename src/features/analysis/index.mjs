import chalk from 'chalk';
import Configstore from 'configstore';
import Debug from 'debug';
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import open from 'open';
import semver from 'semver';
import { callbackServer } from './callback-server.mjs';
import tmp from 'tmp';

import { WEB_APP_URL } from '../../constants.mjs';
import { canReachRepo, downloadRepo, getDefaultBranch } from './github.mjs';
import { GQLClient } from './gql.mjs';
import { githubIntegrationPrompt, mobbAnalysisPrompt } from './prompts.mjs';
import { getSnykReport } from './snyk.mjs';
import { uploadFile } from './upload-file.mjs';
import { keypress, Spinner, CliError } from '../../utils.mjs';
import { pack } from './pack.mjs';
import { getGitInfo } from './git.mjs';

const webLoginUrl = `${WEB_APP_URL}/cli-login`;
const githubSubmitUrl = `${WEB_APP_URL}/gh-callback`;
const githubAuthUrl = `${WEB_APP_URL}/github-auth`;

const MOBB_LOGIN_REQUIRED_MSG = `üîì Login to Mobb is Required, you will be redirected to our login page, once the authorization is complete return to this prompt, ${chalk.bgBlue(
    'press any key to continue'
)};`;
const tmpObj = tmp.dirSync({
    unsafeCleanup: true,
});

const debug = Debug('mobbdev:index');
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const packageJson = JSON.parse(
    fs.readFileSync(path.join(__dirname, '../../../package.json'), 'utf8')
);

if (!semver.satisfies(process.version, packageJson.engines.node)) {
    throw new CliError(
        `${packageJson.name} requires node version ${packageJson.engines.node}, but running ${process.version}.`
    );
}

const config = new Configstore(packageJson.name, { token: '' });
debug('config %o', config);

export async function runAnalysis(
    { repo, scanner, scanFile, apiKey, ci, commitHash, srcPath, ref },
    options
) {
    try {
        await _scan(
            {
                dirname: tmpObj.name,
                repo,
                scanner,
                scanFile,
                ref,
                apiKey,
                ci,
                srcPath,
                commitHash,
            },
            options
        );
    } finally {
        tmpObj.removeCallback();
    }
}

export async function _scan(
    { dirname, repo, scanFile, branch, apiKey, ci, srcPath, commitHash, ref },
    { skipPrompts = false } = {}
) {
    debug('start %s %s', dirname, repo);
    const { createSpinner } = Spinner({ ci });
    skipPrompts = skipPrompts || ci;
    let token = config.get('token');
    debug('token %s', token);
    apiKey ?? debug('token %s', apiKey);
    let gqlClient = new GQLClient(apiKey ? { apiKey } : { token });
    await handleMobbLogin();
    const { projectId, organizationId } = await gqlClient.getOrgAndProjectId();
    const uploadData = await gqlClient.uploadS3BucketInfo();
    let reportPath = scanFile;

    if (srcPath) {
        return await uploadExistingRepo();
    }

    const userInfo = await gqlClient.getUserInfo();
    let { githubToken } = userInfo;
    const isRepoAvailable = await canReachRepo(repo, {
        token: userInfo.githubToken,
    });
    if (!isRepoAvailable) {
        if (ci) {
            throw new Error(
                `Cannot access repo ${repo} with the provided token, please visit ${githubAuthUrl} to refresh your Github token`
            );
        }
        const { token } = await handleGithubIntegration();
        githubToken = token;
    }

    const reference =
        ref ?? (await getDefaultBranch(repo, { token: githubToken }));
    debug('org id %s', organizationId);
    debug('project id %s', projectId);
    debug('default branch %s', reference);

    const repositoryRoot = await downloadRepo(
        {
            repoUrl: repo,
            reference,
            dirname,
            ci,
        },
        { token: githubToken }
    );

    if (!reportPath) {
        reportPath = await getReportFromSnyk();
    }

    await uploadFile(
        reportPath,
        uploadData.url,
        uploadData.uploadKey,
        uploadData.uploadFields
    );
    const mobbSpinner = createSpinner('üïµÔ∏è‚Äç‚ôÇÔ∏è Initiating Mobb analysis').start();
    try {
        await gqlClient.submitVulnerabilityReport(
            uploadData.fixReportId,
            repo,
            reference,
            projectId
        );
    } catch (e) {
        mobbSpinner.error({ text: 'üïµÔ∏è‚Äç‚ôÇÔ∏è Mobb analysis failed' });
        throw e;
    }

    mobbSpinner.success({
        text: 'üïµÔ∏è‚Äç‚ôÇÔ∏è Generating fixes...',
    });

    await askToOpenAnalysis();
    async function getReportFromSnyk() {
        const reportPath = path.join(dirname, 'report.json');

        if (
            !(await getSnykReport(reportPath, repositoryRoot, { skipPrompts }))
        ) {
            debug('snyk code is not enabled');
            return;
        }
        return reportPath;
    }
    async function askToOpenAnalysis() {
        const reportUrl = `${WEB_APP_URL}/organization/${organizationId}/project/${projectId}/report/${uploadData.fixReportId}`;
        !ci && console.log('You can access the report at: \n');
        console.log(reportUrl);
        !skipPrompts && (await mobbAnalysisPrompt());

        !ci && open(reportUrl);
        !ci &&
            console.log(
                chalk.bgBlue(
                    '\n\n  My work here is done for now, see you soon! üïµÔ∏è‚Äç‚ôÇÔ∏è  '
                )
            );
    }

    async function handleMobbLogin() {
        if (
            (token && (await gqlClient.verifyToken())) ||
            (apiKey && (await gqlClient.verifyToken()))
        ) {
            createSpinner().start().success({
                text: 'üîì Logged in to Mobb successfully',
            });

            return;
        }
        if (apiKey && !(await gqlClient.verifyToken())) {
            createSpinner().start().error({
                text: 'üîì Logged in to Mobb failed - check your api-key',
            });
            throw new CliError();
        }
        const mobbLoginSpinner = createSpinner().start();
        if (!skipPrompts) {
            mobbLoginSpinner.update({ text: MOBB_LOGIN_REQUIRED_MSG });
            await keypress();
        }

        mobbLoginSpinner.update({
            text: 'üîì Waiting for Mobb login...',
        });

        const loginResponse = await callbackServer(
            webLoginUrl,
            `${webLoginUrl}?done=true`
        );
        token = loginResponse.token;

        gqlClient = new GQLClient({ token });

        if (!(await gqlClient.verifyToken())) {
            mobbLoginSpinner.error({
                text: 'Something went wrong, API token is invalid.',
            });
            throw new CliError();
        }
        debug('set token %s', token);
        config.set('token', token);
        mobbLoginSpinner.success({ text: 'üîì Login to Mobb successful!' });
    }
    async function handleGithubIntegration() {
        const addGithubIntegration = skipPrompts
            ? true
            : await githubIntegrationPrompt();

        const githubSpinner = createSpinner(
            'üîó Waiting for github integration...'
        ).start();
        if (!addGithubIntegration) {
            githubSpinner.error();
            throw Error('Could not reach github repo');
        }
        const result = await callbackServer(
            githubAuthUrl,
            `${githubSubmitUrl}?done=true`
        );
        githubSpinner.success({ text: 'üîó Github integration successful!' });
        return result;
    }
    async function uploadExistingRepo() {
        const gitInfo = await getGitInfo(srcPath);
        const zipBuffer = await pack(srcPath);

        await uploadFile(
            reportPath,
            uploadData.url,
            uploadData.uploadKey,
            uploadData.uploadFields
        );
        await uploadFile(
            zipBuffer,
            uploadData.repoUrl,
            uploadData.repoUploadKey,
            uploadData.repoUploadFields
        );

        const mobbSpinner = createSpinner(
            'üïµÔ∏è‚Äç‚ôÇÔ∏è Initiating Mobb analysis'
        ).start();

        try {
            await gqlClient.submitVulnerabilityReport(
                uploadData.fixReportId,
                repo || gitInfo.repoUrl,
                branch || gitInfo.reference,
                projectId,
                commitHash || gitInfo.hash
            );
        } catch (e) {
            mobbSpinner.error({ text: 'üïµÔ∏è‚Äç‚ôÇÔ∏è Mobb analysis failed' });
            throw e;
        }

        mobbSpinner.success({
            text: 'üïµÔ∏è‚Äç‚ôÇÔ∏è Generating fixes...',
        });
        await askToOpenAnalysis();
    }
}
