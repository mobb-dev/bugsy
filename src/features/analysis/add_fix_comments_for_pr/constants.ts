import { Scanner } from '@mobb/bugsy/constants'

import { MOBB_ICON_IMG } from '../scm'

export const contactUsMarkdown = `For specific requests [contact us](https://mobb.ai/contact) and we'll do the most to answer your need quickly.`
export const MobbIconMarkdown = `![image](${MOBB_ICON_IMG})`
export const noVulnerabilitiesFoundTitle = `# ${MobbIconMarkdown} No security issues were found âœ…`

export const COMMIT_FIX_SVG = `https://app.mobb.ai/gh-action/commit-button.svg`

export const scannerToFriendlyString: Record<Scanner, string> = {
  checkmarx: 'Checkmarx',
  codeql: 'CodeQL',
  fortify: 'Fortify',
  snyk: 'Snyk',
} as const
