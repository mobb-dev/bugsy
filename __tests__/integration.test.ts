import fs from 'node:fs'
import path from 'node:path'

import open from 'open'
import { expect, it, vi } from 'vitest'

import { API_URL, PROJECT_PAGE_REGEX } from '../src/constants'
import * as analysisExports from '../src/features/analysis'
import {
  CREATE_COMMUNITY_USER,
  PERFORM_CLI_LOGIN,
} from '../src/features/analysis/graphql/mutations'

vi.mock('../src/utils/dirname.ts')

vi.mock('open', () => ({
  default: vi.fn().mockImplementation(async (url: string) => {
    const match = url.match(/\/cli-login\/(.*?)\?.*$/)

    if (match && match.length == 2) {
      const loginId = match[1]

      const createUserRes = await fetch(API_URL, {
        headers: {
          authorization: `Bearer ${process.env.TOKEN}`,
        },
        body: JSON.stringify({
          query: CREATE_COMMUNITY_USER,
        }),
        method: 'POST',
      })
      expect(createUserRes.status).toStrictEqual(200)

      // Emulate "Authenticate" button click in the Web UI.
      const performLoginRes = await fetch(API_URL, {
        headers: {
          authorization: `Bearer ${process.env.TOKEN}`,
        },
        body: JSON.stringify({
          query: PERFORM_CLI_LOGIN,
          variables: {
            loginId,
          },
        }),
        method: 'POST',
      })
      expect(performLoginRes.status).toStrictEqual(200)
    }
  }),
}))

vi.mock('configstore', () => {
  const Configstore = vi.fn()

  Configstore.prototype.get = vi.fn()
  Configstore.prototype.set = vi.fn()
  return { default: Configstore }
})

vi.mock('../src/features/analysis/snyk', () => ({
  getSnykReport: vi.fn().mockImplementation(async (reportPath) => {
    fs.copyFileSync(path.join(__dirname, 'report.json'), reportPath)
    return true
  }),
}))

it('Full analyze flow', async () => {
  open.mockClear()
  const runAnalysisSpy = vi.spyOn(analysisExports, 'runAnalysis')

  await analysisExports.runAnalysis(
    {
      repo: 'https://github.com/mobb-dev/simple-vulnerable-java-project',
      ci: false,
    },
    { skipPrompts: true }
  )

  expect(runAnalysisSpy).toHaveBeenCalled()
  expect(open).toHaveBeenCalledTimes(2)
  expect(open).toBeCalledWith(expect.stringMatching(PROJECT_PAGE_REGEX))
}, 30000)

it('Direct repo upload', async () => {
  open.mockClear()
  await analysisExports.runAnalysis(
    {
      repo: 'https://bitbucket.com/a/b',
      ref: 'test',
      commitHash: 'ad00119b0d4a56f44a49d3d20eccb77978a363f8',
      scanFile: path.join(__dirname, 'assets/simple/codeql_report.json'),
      srcPath: path.join(__dirname, 'assets/simple'),
      ci: false,
    },
    { skipPrompts: true }
  )
  expect(open).toHaveBeenCalledTimes(2)
  expect(open).toBeCalledWith(expect.stringMatching(PROJECT_PAGE_REGEX))
})
