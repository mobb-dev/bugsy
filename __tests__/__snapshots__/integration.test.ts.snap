// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mcp tests > should respond to mcp list_tools call 1`] = `
{
  "tools": [
    {
      "description": "Scans a given local repository for security vulnerabilities and returns auto-generated code fixes.

When to invoke:
• Use when the user explicitly asks to "scan for vulnerabilities", "run a security check", or "test for security issues" in a local repository.
• The repository must exist on disk; supply its absolute path with the required "path" argument.
• Ideal after the user makes code changes (added/modified/staged files) but before committing, or whenever they request a full rescan.

How to invoke:
• Required argument:
  – path (string): absolute path to the repository root.
• Optional arguments:
  – offset (number): pagination offset used when the result set is large.
  – limit (number): maximum number of fixes to include in the response.
  – maxFiles (number): maximum number of files to scan (default: 10). Provide this value to increase the scope of the scan.
  – rescan (boolean): true to force a complete rescan even if cached results exist.

Behaviour:
• If the directory is a valid Git repository, the tool scans the changed files in the repository. If there are no changes, it scans the files included in the las commit.
• If the directory is not a valid Git repository, the tool falls back to scanning recently changed files in the folder.
• If maxFiles is provided, the tool scans the maxFiles most recently changed files in the repository.
• By default, only new, modified, or staged files are scanned; if none are found, it checks recently changed files.
• The tool NEVER commits or pushes changes; it only returns proposed diffs/fixes as text.

Return value:
The response is an object with a single "content" array containing one text element. The text is either:
• A human-readable summary of the fixes / patches, or
• A diagnostic or error message if the scan fails or finds nothing to fix.

Example payload:
{
  "path": "/home/user/my-project",
  "limit": 20,
  "maxFiles": 50,
  "rescan": false
}",
      "display_name": "Scan and Fix Vulnerabilities",
      "inputSchema": {
        "properties": {
          "limit": {
            "description": "[Optional] maximum number of results to return",
            "type": "number",
          },
          "maxFiles": {
            "description": "[Optional] maximum number of files to scan (default: 10). Use higher values for more comprehensive scans or lower values for faster performance.",
            "type": "number",
          },
          "offset": {
            "description": "[Optional] offset for pagination",
            "type": "number",
          },
          "path": {
            "description": "Full local path to repository to scan and fix vulnerabilities",
            "type": "string",
          },
          "rescan": {
            "description": "[Optional] whether to rescan the repository",
            "type": "boolean",
          },
        },
        "required": [
          "path",
        ],
        "type": "object",
      },
      "name": "scan_and_fix_vulnerabilities",
    },
    {
      "description": "Check the MOBB backend for pre-generated fixes (patch sets) that correspond to vulnerabilities detected in the supplied Git repository.

Use when:
• You already have a local clone of a Git repository and want to know if MOBB has fixes available for it.
• A vulnerability scan has been run previously and uploaded to the MOBB backend and you want to fetch the list or count of ready-to-apply fixes before triggering a full scan-and-fix flow.

Required argument:
• path – absolute path to the local Git repository clone.

Optional arguments:
• offset – pagination offset (integer).
• limit  – maximum number of fixes to return (integer).

The tool will:
1. Validate that the provided path is secure and exists.
2. Verify that the directory is a valid Git repository with an "origin" remote.
3. Query the MOBB service by the origin remote URL and return a textual summary of available fixes (total and by severity) or a message if none are found.

Call this tool instead of scan_and_fix_vulnerabilities when you only need a fixes summary and do NOT want to perform scanning or code modifications.",
      "display_name": "Fetch Available Fixes",
      "inputSchema": {
        "properties": {
          "limit": {
            "description": "[Optional] maximum number of results to return",
            "type": "number",
          },
          "offset": {
            "description": "[Optional] offset for pagination",
            "type": "number",
          },
          "path": {
            "description": "Full local path to the cloned git repository to check for available fixes",
            "type": "string",
          },
        },
        "required": [
          "path",
        ],
        "type": "object",
      },
      "name": "fetch_available_fixes",
    },
    {
      "description": "Continuesly monitors your code and scans for new security vulnerabilities.

When to invoke:
• ALWAYS call this tool ONCE at the very end of every reasoning / code-generation session (after all code edits, linting, tests, etc.).
• Invoke when the user explicitly asks to "check for new fixes", "look for fresh patches", or similar.
• The repository must exist locally and be a valid Git repo with an 'origin' remote configured.

How to invoke:
• Required argument:
  – path (string): absolute path to the repository root.

Behaviour:
• If no new fixes are available, it returns a concise message indicating so.
• If fixes are found, it returns a human-readable summary including total count and severity breakdown.

Example payload:
{
  "path": "/home/user/my-project"
}",
      "display_name": "Check for New Available Fixes",
      "inputSchema": {
        "properties": {
          "path": {
            "description": "Full local path to the cloned git repository to check for new available fixes",
            "type": "string",
          },
        },
        "required": [
          "path",
        ],
        "type": "object",
      },
      "name": "check_for_new_available_fixes",
    },
  ],
}
`;
